src/Raven.Server/Documents/Sharding/Handlers/ShardedIncomingReplicationHandler.cs

```diff
@@ -140,8 +166,101 @@ internal async Task<(string MergedChangeVector, long LastEtag)> GetInitialHandsh
+        }
+
+        protected override void OnDocumentsReceived()
+        {
+            if (_shardedSource == false)
+                return;
+
+            var states = ShardReplicationLoader.GetShardedExternalReplicationStates(_parent.Server, _parent.DatabaseName, _sourceShardedDatabaseName, _sourceShardedDatabaseId);
+
+            states ??= new ShardedExternalReplicationState
+            {
+                NodeTag = _parent._server.NodeTag,
+                SourceDatabaseName = _sourceShardedDatabaseName,
+                SourceShardedDatabaseId = _sourceShardedDatabaseId,
+                ReplicationStates = new Dictionary<string, ShardedExternalReplicationStateForSingleSource>()
+            };
+
+            bool update = false;
+            foreach (var (shardNumber, handler) in _handlers)
+            {
+                if (_lastSentEtagPerDestination[shardNumber] == 0)
+                    continue;
+
+                if (states.ReplicationStates.TryGetValue(ConnectionInfo.SourceDatabaseName, out var shardedSourceState) &&
+                    shardedSourceState.DestinationStates.TryGetValue(handler.DatabaseName, out var shardedDestinationState))
+                {
+                    if (_lastSentEtagPerDestination[shardNumber] == shardedDestinationState.LastSentEtag)
+                        continue;
+
+                    // update existing value
+                    shardedDestinationState.LastSentEtag = _lastSentEtagPerDestination[shardNumber];
+                    shardedDestinationState.DestinationChangeVector = handler.LastAcceptedChangeVector;
+                }
+                else
+                {
+                    shardedDestinationState = new ExternalReplicationState
+                    {
+                        NodeTag = _parent._server.NodeTag,
+                        LastSentEtag = _lastSentEtagPerDestination[shardNumber],
+                        DestinationChangeVector = handler.LastAcceptedChangeVector
+                    };
+
+                    shardedSourceState ??= new ShardedExternalReplicationStateForSingleSource();
+                    shardedSourceState.DestinationStates ??= new Dictionary<string, ExternalReplicationState>();
+                }
+
+                update = true;
+                shardedSourceState.DestinationStates[handler.DatabaseName] = shardedDestinationState;
+                shardedSourceState.LastSourceEtag = _lastDocumentEtag;
+                shardedSourceState.LastSourceChangeVector = _lastAcceptedChangeVectorDuringHeartbeat;
+                states.ReplicationStates[ConnectionInfo.SourceDatabaseName] = shardedSourceState;
+            }
+
+            if (update == false)
+                return;
+
+            var command = new ShardedUpdateExternalReplicationStateCommand(_parent.DatabaseName, RaftIdGenerator.NewId())
+            {
+                ReplicationState = states
+            };
+
+            _parent._server.SendToLeaderAsync(command)
```

Comment:
especially that you aren't waiting here.. it is possible for you to overwrite new data with old data

Reply:
done